### ISSUE TITLE
Add `pip list --outdated`

### ISSUE THREAD

**Issue by pfmoore** [astral-sh/uv#2150]
`pip list --outdated` is also a useful feature.

_Originally posted by @ismail in https://github.com/astral-sh/uv/issues/1401#issuecomment-1947899132_

Creating a separate issue for visibility, as the issue for creating the `uv pip list` command has now been closed as complete. Apologies if I missed an issue that already requested this - feel free to close this as a duplicate if that's the case.

I use `pip list --outdated` all the time (more than any other form of `pip list`, probably). It would be really useful to have in `uv` as well.

**Comment by timleslie** [comment]
@zanieb @charliermarsh I was wondering if there were any plans for adding this feature in the coming releases? This is one of the last pieces of the puzzle for our full transition to `uv` from `pip`. Any kind of ball-park estimate of if/when would help with our internal planning 🙏

**Comment by zanieb** [comment]
We don't have any immediate plans to implement this although it seems like it wouldn't be particularly challenging (we have all the components needed for this).

Can you share more about your use-case?

I'd be happy to review a contribution adding support for this.

**Comment by timleslie** [comment]
Thanks for the reply. We have a somewhat complicated dependency management process which precludes us from using tools like renovate or dependabot on our repo. As such we use `pip list --outdated --format=json` together with some custom postprocessing of the output and our `pyproject.toml` to create a list of updatable packages.

From that point of view this is more of a "nice to have" rather than a show-stopper, but for our team this functionality gives us a big quality of life improvement in our DX.

I'm happy to take a swing at the implementation if all the pieces are there 🤞

**Comment by timleslie** [comment]
(And while I have you, many thanks for embarking on this project. `pip` install times have been a thorn in our side for an eternity, and `uv` is an absolute game changer for us 🙏)

**Comment by pfmoore** [comment]
I’d support this. I use `pip list --outdated` regularly to check what needs updating in an environment. There isn’t really a good alternative in my view (if there is, I’d be interested to know what it is).

**Comment by zanieb** [comment]
Thanks for all the kind words :) that's an interesting use-case but it makes some sense to me. I imagine it'd make sense to have a future workflow that does not depend on listing the installed packages and can just give you metadata directly from a lockfile... but we can think more about that separately.

Feel free to give it a try. I'm happy to provide any guidance. I'd look at the `pip_install` implementation and the `Plan.reinstalls` field.

**Comment by timleslie** [comment]
Sounds good, I'll try to chip away at this over the coming days and see where I get to. So far I've got a build going and added the CLI arg, so I now I just need to fill in:

```rust
    if outdated {
        ...
    }
```

Simple :-) Just getting familiar with the resolver structs and how all the pieces fit together. I'll reach out with any questions and ship a draft PR for feedback when I have something to look at. `pip_install` and `Plan.reinstalls` sound like good pointers 👍

**Comment by timleslie** [comment]
@zanieb I've pushed up a WIP branch as #2582 which explores my first attempt. I'd appreciate any feedback on whether this is an appropriate direction, or whether there's an alternate path that might involve less noisy code 🤔

**Comment by gmankab** [comment]
i am writing autotests, and one of them is checking for oudated dependencies

i use `pip list --outdated --format=json`

and it is sooo slow

i hope uv will be much faster

waiting for the pull request to be accepted

**Comment by AndydeCleyre** [comment]
FWIW I think that the lack of `pip list -o` and `pip list --pre` are the last two items I need to replace all usage of pip and pip-tools with `uv` (when installed) in my project, [zpy](https://github.com/AndydeCleyre/zpy/issues/25). Although if `uv` always acts like `--pre`, that would be fine as well.

**Comment by mralusw** [comment]
How can one upgrade all packages in a venv without this command (or `pip freeze --local`, which also isn't supported)? Isn't this a really common task? It's been amazingly hard, for many years now, to do this with python, but without these two commands in `uv` one has to crawl back to slow `pip`.

**Comment by AndydeCleyre** [comment]
@mralusw You can keep track of the packages in requirements+lock files, upgrade those, then sync.

```console
$ uv pip compile -U requirements.in -o requirements.txt
$ uv pip sync requirements.txt
```

Or without any locking or syncing, just listing top level pkgs, this may do: 

```console
$ uv pip install -Ur requirements.in
```

Otherwise you might collect the names from

```console
$ uv pip list --format json
```

then `uv pip install -U ` them.

**Comment by T-256** [comment]
Just an opinion at here:
use it as separated command, like `pip update`, which takes `requirements.txt` (compiled versions `{pkg}=={version}`) as input and output is updated list.
for current venv and its installed packages in it we can then use `uv pip freeze | uv pip update`.

Also, fwiw I opened a enhancement request for `uv pip freeze` in https://github.com/astral-sh/uv/issues/3141

**Comment by gmankab** [comment]
hello, is there any progress with implementing this?

**Comment by mralusw** [comment]
While there are work-arounds, note that `pip` can be used not only from the command line, but also from scripts and higher-order software. The update-check is conceptually different from the update-install step.

For example, when writing a package manager, it's a natural workflow to check for available updates, notify the user, allow them to select an action, and then separately perform the update. Only the last step might need admin privileges, and only the last two need an interactive environment.

Also, it's a waste to have a resolution mechanism and not expose it (especially when there's a clear API, as in this case).

**Comment by NeilGirdhar** [comment]
@zanieb 
> Can you share more about your use-case?

I think this is a really good question and illustrates your team's desire to provide good interfaces.

For me, I use this in a few places:

* As a faster version of `poetry update --dry-run`.  Of course, if the Poetry command were faster, then that would be preferable since it would give more accurate results.
* After `poetry update` to find out if anything wasn't upgraded, and then I do `poetry show -t --ansi | less`  to find out why certain things weren't updated.
* In my system-wide Python to find out if anything needs updating.  For me, I would be okay with my system-wide Python being managed in the same way as a Poetry project—if that were possible.

Thus, I don't really need `uv pip list --outdated`.  What I need are fast versions of:
* `uv poetry update --dry-run`,
* `uv poetry show -o`, and
* some way to manage my system Python libraries as if they were in a Poetry project.

**Comment by mralusw** [comment]
Just curious, is there a technical reason why the results of the package resolution process might be difficult to expose?

I'm writing a package manager that sits on top of `pip` and other package sources. I'm trying to avoid fancy tools (see for instance [Why not tell people to "simply" use pyenv, poetry, pipx or anaconda](https://www.bitecode.dev/p/why-not-tell-people-to-simply-use) and the related series for why sticking to upstream python packaging tools might avoid a host of problems). `uv pip` is not of course vanilla `pip`, but it's close, doesn't require custom workflows, and the speed is compelling.

**Comment by zanieb** [comment]
I don't think it's hard to expose the package resolution results per-say, it's just a good bit of work and we're not prioritizing adding output formats right now. See https://github.com/astral-sh/uv/issues/3199 which is a bit more focused on the goal of machine readable output.

**Comment by adiberk** [comment]
@zanieb 
Just to throw in my two cents here - I apologize if this isn't applicable, but at my company we needed to upgrade to new python version and wanted to update our dependencies in our pyproject to the latest possible version
A command like `poetry update --latest` (using this plugin https://github.com/MousaZeidBaker/poetry-plugin-up)
Was extremely helpful and useful.

Would something like this be on the road map for uv?  Though we use dependabot now to help keep our dependencies up to date, I believe this functionality is now viewed as critical by our eng team.

**Comment by ziddey** [comment]
In case it helps someone, this is how I've been simulating `pip list --outdated`:
```
bash -c "uv pip list --format=freeze |sed 's/==.*//' | uv pip compile - --no-deps --no-header |diff <(uv pip list --format=freeze) - -y --suppress-common-lines || :"
```
Obviously not in the same format, but it serves to show held back packages.

**Comment by gh640** [comment]
I also encountered this issue, so for the time being, I decided to run `uv sync --seed` to install pip in my venv and then run `uv run pip list --outdated`.

I hope this will be helpful to someone.

**Comment by charliermarsh** [comment]
Bun has this now too, it's nice!

![Screenshot 2024-08-30 at 10 57 57 AM](https://github.com/user-attachments/assets/ce89a628-0973-418e-bad4-4e01d16aa74c)

**Comment by dwt** [comment]
Another use case: I use `pip list --outdated` regularly when coming back to a project after some time, to quickly see what work I have ahead of me.

**Comment by golgor** [comment]
@zanieb 
The use-case we have is keeping our IoT Pipeline up to date and to avoid to large required changes or tech backlog.

For example, every time I work with a PR, I can just run one command to check the status of all the used packages. If I see that quite a few packages that could be updated, I would start looking into this. I do it like this using Poetry and it works very good with keeping things up to date.

The alternative would be to leave this to random chance (check when you remember to do it) or set up something like a quartely calendar event to check it.

It would also be a neater way to automate this with for example a Github Action (which is possible with the script from @ziddey, but would be better with a an official solution).

**Comment by NeilGirdhar** [comment]
> Another use case: I use pip list --outdated regularly when coming back to a project after some time, to quickly see what work I have ahead of me.

Wouldn't this be better as `uv tree --outdated`?   After all, it doesn't matter which dependencies have updates, but rather which dependencies _can be_ updated.

**Comment by AndydeCleyre** [comment]
I don't know about a tree output, which may be nice, but FWIW regarding output format: I currently wrap `pip --disable-pip-version-check list -o --format json` with a function that presents a table for me of outdated packages across an arbitrary number of projects (venvs).

**Comment by NeilGirdhar** [comment]
My point wasn't that the output format.  My point was that the use case suggested above should be showing the packages that can be updated giving the lock rather than simply the outdated packages.

**Comment by dwt** [comment]
This seems sensible - coming from pip, `pip list --outdated` already shows all available updates, not just the direct dependencies. Is there any reason why `uv pip list --outdated` would only show outdated direct dependencies? Or do you just mean that the format would be nicer from `uv tree --outdated`?

**Comment by NeilGirdhar** [comment]
> This seems sensible - coming from pip, pip list --outdated already shows all available updates, not just the direct dependencies. 

Sorry, I think we're talking past each other.  As I said in my last comment, this has nothing to do with format.  My point was that the use case suggested above should be showing **the packages that can be updated given the lock** rather than **simply the outdated packages**.

Consider, for example that you have dependencies:
```
dependencies = [
    'jaxlib >=0.4.14, <=0.4.31',
    'numpy <= 2.1.0'
    ...
```
And you have jaxlib==0.4.31 and numpy==2.0.0 installed.  Suppose that jaxlib 0.5.0 and numpy 2.1.0 are available. What should be displayed?

pip list --outdated naively says: jaxlib 0.5.0 and numpy 2.1.0

That's because it doesn't see your pyproject.toml file.  I think it's more useful (considering the use cases mentioned in this thread, which are project use cases) to show only that numpy is outdated.  This is because `uv lock -U` will not update jaxlib, and it will be both useless and confusing (in my opinion)  to keep saying that it is outdated and refuse to update it.

**Comment by dimbleby** [comment]
> This is what poetry show -o currently does.

no it isn't. poetry shows all packages for which newer versions are available even if `poetry lock` would not use those newer versions.

**Comment by NeilGirdhar** [comment]
> fwiw I find this behaviour more useful than telling me which packages it would upgrade if I re-locked. (I can `update --dry-run` for

Ah, my mistake.  For some reason I thought it was doing the dry run essentially.

> whereas telling me that I have an upper bound that I might want to relax is something I want to know.

For me, I get a lot of noise from downstream package limits that I can't relax holding things back.  That's why I'd rather just see what I can control.  Maybe it would be nice to show at least two sections:
* things that would be upgraded given `uv lock -U`, and
* things that could be upgraded if constraints in the project file were relaxed.

This would still hide some outdated packages if they are blocked by dependencies within dependencies.  WDYT?

**Comment by dwt** [comment]
Sounds sensible, those are three use cases. The last one would be the least useful to me, but the first two I use very regularly.

**Comment by ericbn** [comment]
Looks like there are different features being discussed here. I propose we use this thread to only discuss what seems to be the original feature, which is what `pip list --outdated` does and also what the script posted [above](https://github.com/astral-sh/uv/issues/2150#issuecomment-2313934453) by @ziddey does, which is: to list outdated packages **based on which packages are installed on the current virtual environment (or in the system)**. This ignores any lock files or pyproject.toml files, as it only looks at which packages are currently installed.

**Comment by RatulMaharaj** [comment]
I would love for this to be a thing. I mostly use `pnpm` for my node.js projects and to me `uv` feels super familiar because of that.

In that ecosystem, my favourite and most commonly used commands are `pnpm add`, `pnpm remove` and `pnpm outdated`. We already have `uv add` and `uv remove` - I think being able to run `uv outdated` would be awesome too!

**Comment by rafalkrupinski** [comment]
Not sure what the command is supposed to do but with this in project.dependencies:

"boltons>=24.1.0",

it doesn't show that boltons is outdated (25.0.0 available)

**Comment by AndydeCleyre** [comment]
AFAIU, like pip, it's reporting on the actually installed version vs the newest release, disregarding stated requirements.

**Comment by zanieb** [comment]
Yeah this is a `pip` command which is not project-aware, it's reading the environment

```
❯ uv venv
Using CPython 3.13.0
Creating virtual environment at: .venv
Activate with: source .venv/bin/activate
❯ uv pip install boltons==24.1.0
Resolved 1 package in 128ms
Prepared 1 package in 65ms
Installed 1 package in 3ms
 + boltons==24.1.0
❯ uv pip list --outdated
Package Version Latest Type
------- ------- ------ -----
boltons 24.1.0  25.0.0 wheel
```

**Comment by htran-ubed** [comment]
Could this be added to the documentation of managing packages or maybe somewhere else? As I could not find it anywhere in the documentation.

**Comment by zanieb** [comment]
@htran-ubed we tend not to spend too much time on documenting `pip` command since it's just a parity interface. Feel free to open a pull request adding it to the documentation though.

**Comment by AndydeCleyre** [comment]
For anyone documenting, note that it's not an _exact_ parity interface, e.g.:

```console
$ pip list --outdated --format columns --exclude-editable --format json --include-editable     # valid
$ uv pip list --outdated --format columns --exclude-editable --format json --include-editable  # invalid for two reasons
```

**Comment by tyilo** [comment]
> Im using
> 
> uv tree --outdated --depth 1

This doesn't seem to work anymore in uv 0.9.4. It just lists all dependencies.

`uv pip list --outdated` works instead.

**Comment by keikoro** [comment]
> > Im using
> > uv tree --outdated --depth 1
> 
> This doesn't seem to work anymore in uv 0.9.4. It just lists all dependencies.

It works in the latest release, [0.9.26](https://github.com/astral-sh/uv/releases/tag/0.9.26).

**Comment by keikoro** [comment]
... However, the two commands aren't interchangeable, actually.

Only the pip interface will help determine the status of _**installed packages**_:

```shell
$ uv pip list --outdated
```
... which is what this issue was originally opened for.

`uv tree` visualises the _**project dependencies**_; i.e. this highlights what's outdated in `pyproject.toml` files:

```shell
$ uv tree --outdated --depth 1
```

The latter command won't e.g. let you know about manually installed editable packages which are outdated (dependencies added via `uv pip install -e <local project path>`).

**Comment by tyilo** [comment]
> > > Im using
> > > uv tree --outdated --depth 1
> > 
> > 
> > This doesn't seem to work anymore in uv 0.9.4. It just lists all dependencies.
> 
> It works in the latest release, [0.9.26](https://github.com/astral-sh/uv/releases/tag/0.9.26).

Hmm, I guess it does work, but not in the way I expected.

```
uv tree --depth 1
```
and
```
uv tree --depth 1 --outdated
```
lists the same packages.

You need to use something like
```
uv tree --depth 1 --outdated | grep '(latest: '
```
to only get the list of outdated packages.

**Comment by keikoro** [comment]
> You need to use something like
> 
> ```
> uv tree --depth 1 --outdated | grep '(latest: '
> ```
> to only get the list of outdated packages.

@tyilo You can also just grep for "latest". No need to add the surrounding special characters.

**Comment by tyilo** [comment]
> [@tyilo](https://github.com/tyilo) You can also just grep for "latest". No need to add the surrounding special characters.

Not if any of my dependencies contains `latest` in the name.

**Comment by keikoro** [comment]
> Not if any of my dependencies contains `latest` in the name.

Oh, good point!
